solidity/solidity-interfaces.md

```diff
@@ -0,0 +1,84 @@
+# Updated Solidity Interface Policy
+
+## Context
+
+Smart contract interfaces are useful tools that allow external tools, contracts, and tests to
+interact with smart contracts using their ABIs instead of interacting with the underlying
+implementation contract. Interfaces are generally more flexible than implementations because we
+pin implementations to specific compiler versions whereas interfaces can have carat versions that
+cover an entire Solidity major version.
+
+Interfaces are also valuable because they create a separation between source contracts and other
+contracts that need to interact with those source contracts. This separation significantly
+decreases compilation time when changes need to be made to source contracts. Any changes to a
+source contract would otherwise trigger a re-compilation of many other contracts that rely on that
+source contract, resulting in untolerably long waits.
+
+## Problem Statement
+
+We recently undertook an update to the Optimism Monorepo that added interfaces for all smart
+contracts. This resulted in a significant improvement to compilation time but simultaneously added
+complexity for any developer that needs to introduce a new smart contract. Most of this complexity
+arises because existing interface policy is non-obvious and the process for generating interfaces
+is entirely manual. We would like to make certain changes to this policy to support a better
+development process.
+
+We wanted to entirely automate the interface generation process so that users do not need to
+generate interfaces themselves. In exploratory work on this interface generation process, we've
+found that there are a number of important edge cases that came up.
+
+Specifically:
+
+- Source contracts using interfaces creates a basic dependency problem. For instance, suppose that
+  you'd like to add a new contract `ContractA`. You'd also like to create a new contract
+  `ContractB` that depends on `ContractA`. Our current policy states that `ContractB` should use
+  the interface for `ContractA`, `IContractA`. You would therefore need to generate the interface
+  for `ContractA` *before* starting to work on `ContractB`, which is annoying.
+- Circular dependencies in source contracts are basically impossible to resolve without manually
+  crafting contract interfaces. For instance, if `ContractA` depends on `IContractB` and
+  `ContractB` depends on `IContractA` but neither contract interface exists yet (new contracts)
+  then the contracts cannot be compiled (because neither interface exists) and the interfaces can
+  not be generated. Catch 22.
```

Comment:
IMO the two items above can be fairly easily resolved during development just by manually defining the interface being used in the same file as the contract using it, and when the interface is stable it can be autogenerated into `/interfaces`. 

My concern is that creating a new contract is a rare occurrence vs. compiling the source contracts, maybe something on the order of 1:1000 in terms of compilation cycles, thus even a fairly small slow down in the compilation process requires a strong justification.

Reply:
Yeah fair, makes total sense!